isomerpages/isomer-indirection
# Feat(Indirection layer): Create DNS records on push to main branch

## Problem

Isomer does not have an Indirection layer at the moment to add in DNS records programatically. 
Closes [IS-48](https://isomeropengov.atlassian.net/jira/software/projects/IS/boards/1/roadmap?selectedIssue=IS-48).

## Solution
This repo uses Pulumi to add in DNS records in our Cloudflare account. Using Github actions, we are able to add new DNS records on push to main. 

Engineering decisions made:

1) Usage of Pulumi SaaS rather than hosting our own S3 buckets for deployment. 

You may think that Pulumi stacks are like the [backend of of Pulumi](https://www.pulumi.com/docs/intro/concepts/state/). Our stack stores the state of our resources, and the [Pulumi Engine makes a decision on which resources to delete or add based on the changes done on the stack](https://www.pulumi.com/docs/intro/concepts/how-pulumi-works/). Since the [individual account](https://www.pulumi.com/pricing/) is more than able to support this use case, the decision is made to leverage on Pulumi SaaS to store our stacks rather than storing them in a self hosted S3 buckets. In the future, we can also use their secrets service rather than using KMS. 

2) Dynamic imports in `index.ts` file. 

While this is not the conventional method of importing files, I am inclined to use dynamic imports rather than static ones since we are then able to add in just one ts file when we want to add in DNS records for a particular agency. If we have static imports, we would have to keep updating index.ts, which seems like redundant work, and more prone to errors. 

3) Usage of yarn

While we are not moving to yarn in frontend or backend yet, since it is going to be done in the future (correct me if I am wrong) am preemptively using yarn rather than node. 

4) No usage of convict

While I understand that convict is being used to check if env vars exist during run time for our BE and FE codes, I felt this solution would be an overkill here since we only have 1 secret variable for referenced in the file (`CLOUDFLARE_ZONE_ID`), and 3 (`CLOUDFLARE_ZONE_ID`, `CLOUDFLARE_API_TOKEN`, `PULUMI_ACCESS_TOKEN`) that we are storing in github secrets for this repo. I foresee that these are unlikely to change in the near future. If reviewers feel otherwise, am ok to tackle this in another PR.  

5) File structure

The way I envision this going to be similar to the way our redirection repo works. Anytime we want to add DNS records for an agency, we navigate to `/dns` folder and put in one file there. (eg. click [here](https://github.com/isomerpages/isomer-indirection/commit/039571d851651779081a0d3811c3ac81b5df65da#diff-f480af56a0e94dbfd0313de76bffb03fe0c811ee1df12bea3f24be5b89f33159)) This way, if something goes wrong, we can manually just add/delete in files and push to main. (Ops can even do this in the interim while eng builds out the feature). 

## Tests
To aid with testing, I have created a separate branch [here](https://github.com/isomerpages/isomer-indirection/tree/test/github-actions-integration). The code is modified to run github actions on push to the branch and run on a separate stack to not pollute our main production stack. Logging into our cloudflare account will show the `kishoreindirectiontest3.isomer.gov.sg` as a DNS record. 

If you wish to do your own testing: 
1) Add in a file (eg. blah.isomer.gov.sg) with the syntax described in the README.md. An example would be this file [here](https://github.com/isomerpages/isomer-indirection/commit/039571d851651779081a0d3811c3ac81b5df65da#diff-f480af56a0e94dbfd0313de76bffb03fe0c811ee1df12bea3f24be5b89f33159)
2) Push new file to origin
3) Verify that github actions is a success 
4) Verify new DNS record updated in Cloudflare
5) Delete newly added file in step 1 in github locally
6) Push changes to origin

NOTE: Do **_NOT_** delete the DNS record directly in Cloudflare! This will cause the stack and actual infra to be out of sync, leading to errors. 

You can feel free to push to this branch (the github actions is set to run on push to this branch, on a separate stack). 
## Deploy Notes




**New environment variables**:
- `CLOUDFLARE_ZONE_ID`

**New environment variables in Github Secrets** 
- `CLOUDFLARE_API_TOKEN`
- `PULUMI_ACCESS_TOKEN`

The 3 env vars above have been updated in 1pw.

[IS-48]: https://isomeropengov.atlassian.net/browse/IS-48?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
